packages/eslint-plugin/src/rules/contextual-lifecycle.ts

```diff
@@ -0,0 +1,103 @@
+import { TSESTree } from '@typescript-eslint/experimental-utils';
+import { createESLintRule } from '../utils/create-eslint-rule';
+
+import {
+  ANGULAR_CLASS_DECORATOR_LIFECYCLE_METHOD_MAPPER,
+  AngularClassDecorators,
+  AngularLifecycleMethodKeys,
+  getClassName,
+  getDeclaredMethods,
+  getDecorator,
+  getMethodName,
+  isAngularLifecycleMethod,
+} from '../utils/utils';
+
+type Options = [];
+export type MessageIds = 'contextuaLifecycle';
+export const RULE_NAME = 'contextual-lifecycle';
+
+export default createESLintRule<Options, MessageIds>({
+  name: RULE_NAME,
+  meta: {
+    type: 'problem',
+    docs: {
+      description:
+        'Ensures that classes use allowed lifecycle method in its body',
+      category: 'Possible Errors',
+      recommended: false,
+    },
+    schema: [],
+    messages: {
+      contextuaLifecycle: `The method {{methodName}} is not allowed for class {{className}} because it is decorated with {{decoratorName}}`,
+    },
+  },
+  defaultOptions: [],
+  create(context) {
+    return {
+      ClassDeclaration(node: TSESTree.ClassDeclaration) {
```

Comment:
Rather than accessing a single node in the tree and then manually asserting things based on what "kind" it is in a single handler function, usually we want to leverage the fact that we can provide any number of node selectors to ESLint and it will call the dedicated function appropriately:

```ts
import { PIPE_CLASS_DECORATOR, DIRECTIVE_CLASS_DECORATOR } from '../utils/selectors';

// ...etc

return {
  [PIPE_CLASS_DECORATOR]: pipeHandler,
  [DIRECTIVE_CLASS_DECORATOR]: directiveHandler,
  // ... etc
```

They can of course share any number of functions between them so code does not need to be repeated any more than it does in your example